airbyte-integrations/connectors/source-google-ads/source_google_ads/custom_query_stream.py

```diff
@@ -0,0 +1,158 @@
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+
+import re
+from functools import lru_cache
+from typing import Any, Dict, List, Mapping
+
+from .streams import IncrementalGoogleAdsStream
+
+
+class CustomQuery(IncrementalGoogleAdsStream):
+    def __init__(self, custom_query_config, **kwargs):
+        self.custom_query_config = custom_query_config
+        self.user_defined_query = custom_query_config["query"]
+        super().__init__(**kwargs)
+
+    @property
+    def primary_key(self) -> str:
+        """
+        The primary_key option is disabled. Config should not provide the primary key.
+        It will be ignored if provided.
+        If you need to enable it, uncomment the next line instead of `return None` and modify your config
+        """
+        # return self.custom_query_config.get("primary_key") or None
+        return None
+
+    @property
+    def name(self):
+        return self.custom_query_config["table_name"]
+
+    def get_query(self, stream_slice: Mapping[str, Any] = None) -> str:
+        start_date, end_date = self.get_date_params(stream_slice, self.cursor_field)
+        return self.insert_segments_date_expr(self.user_defined_query, start_date, end_date)
+
+    # IncrementalGoogleAdsStream uses get_json_schema a lot while parsing
+    # responses, caching plaing crucial role for performance here.
+    @lru_cache()
+    def get_json_schema(self) -> Dict[str, Any]:
+        """
+        Compose json schema based on user defined query.
+        :return Dict object representing jsonschema
+        """
+
+        local_json_schema = {
+            "$schema": "http://json-schema.org/draft-07/schema#",
+            "type": "object",
+            "properties": {},
+            "additionalProperties": True,
+        }
+        # full list {'ENUM', 'STRING', 'DATE', 'DOUBLE', 'RESOURCE_NAME', 'INT32', 'INT64', 'BOOLEAN', 'MESSAGE'}
+
+        google_datatype_mapping = {
+            "INT64": "integer",
+            "INT32": "integer",
+            "DOUBLE": "number",
+            "STRING": "string",
+            "BOOLEAN": "boolean",
+            "DATE": "string",
+        }
+        fields = CustomQuery.get_query_fields(self.user_defined_query)
+        fields.append(self.cursor_field)
+        google_schema = self.google_ads_client.get_fields_metadata(fields)
+
+        for field in fields:
+            node = google_schema.get(field)
+            # Data type return in enum format: "GoogleAdsFieldDataType.<data_type>"
+            google_data_type = str(node.data_type).replace("GoogleAdsFieldDataType.", "")
+            if google_data_type == "ENUM":
+                field_value = {"type": "string", "enum": list(node.enum_values)}
+            elif google_data_type == "MESSAGE":  # this can be anything (or skip as additionalproperties) ?
+                output_type = ["string", "number", "array", "object", "boolean", "null"]
```

Comment:
let's just make it a string and cast it as that

Reply:
Done, but it took some time,  also added test for stream that contains MESSAGE field.